src/server/pfs/server/api_server_new_storage.go

```diff
@@ -10,35 +10,57 @@ import (
 	"github.com/pachyderm/pachyderm/src/client/pfs"
 	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
 	"github.com/pachyderm/pachyderm/src/client/pkg/grpcutil"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset"
 	"github.com/pachyderm/pachyderm/src/server/pkg/storage/metrics"
 )
 
 func (a *apiServer) PutTar(server pfs.API_PutTarServer) (retErr error) {
-	request, err := server.Recv()
-	func() { a.Log(request, nil, nil, 0) }()
-	defer func(start time.Time) { a.Log(request, nil, retErr, time.Since(start)) }(time.Now())
+	req, err := server.Recv()
+	func() { a.Log(req, nil, nil, 0) }()
+	defer func(start time.Time) { a.Log(req, nil, retErr, time.Since(start)) }(time.Now())
 	return metrics.ReportRequestWithThroughput(func() (int64, error) {
 		if err != nil {
 			return 0, err
 		}
 		if !a.env.NewStorageLayer {
 			return 0, errors.Errorf("new storage layer disabled")
 		}
-		ptr := &putTarReader{
-			server: server,
-			r:      bytes.NewReader(request.Data),
-		}
-		request.Data = nil
-		repo := request.Commit.Repo.Name
-		commit := request.Commit.ID
-		if err := a.driver.putFilesNewStorageLayer(server.Context(), repo, commit, ptr); err != nil {
-			return ptr.bytesRead, err
+		repo := req.Commit.Repo.Name
+		commit := req.Commit.ID
+		var bytesRead int64
+		if err := a.driver.withFileSet(server.Context(), repo, commit, func(fs *fileset.FileSet) error {
+			for {
+				req, err := server.Recv()
+				if err != nil {
+					if err == io.EOF {
+						return nil
+					}
+					return err
+				}
+				ptr := &putTarReader{
+					server: server,
+					r:      bytes.NewReader(req.Data),
+				}
+				defer func() {
+					bytesRead += ptr.bytesRead
+				}()
```

Comment:
There might be ~thousands of these defers - does that have any implications on golang?

Reply:
We should not be computing the bytes read like this. I for the most part pulled the existing put files code over. I'm pretty sure I would have noticed this when reviewing / testing this. My main goal with the current state of this PR was to give you a chance to take a look at the API and see if it is going to work for you.